test/src/test/java/org/corfudb/infrastructure/logreplication/LogEntryWriterTest.java

```diff
@@ -0,0 +1,320 @@
+import org.corfudb.protocols.logprotocol.OpaqueEntry;
+import org.corfudb.protocols.service.CorfuProtocolLogReplication;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.LogReplication;
+import org.corfudb.runtime.LogReplication.LogReplicationEntryMsg;
+import org.corfudb.runtime.LogReplication.LogReplicationEntryMetadataMsg;
+import org.corfudb.runtime.collections.TxnContext;
+import org.corfudb.runtime.view.AbstractViewTest;
+import org.corfudb.runtime.view.Address;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.protobuf.UnsafeByteOperations.unsafeWrap;
+import static org.mockito.Mockito.times;
+
+@SuppressWarnings("checkstyle:magicnumber")
+public class LogEntryWriterTest extends AbstractViewTest {
+
+    private CorfuRuntime corfuRuntime;
+    private LogReplicationMetadataManager metadataManager;
+    private LogEntryWriter logEntryWriter;
+    private TxnContext txnContext;
+
+    @Before
+    public void setUp() {
+        corfuRuntime = getDefaultRuntime();
+        LogReplicationConfig replicationConfig = Mockito.mock(LogReplicationConfig.class);
+        metadataManager = Mockito.mock(LogReplicationMetadataManager.class);
+        Mockito.doReturn(Address.NON_ADDRESS).when(metadataManager).getLastAppliedSnapshotTimestamp();
+        Mockito.doReturn(Address.NON_ADDRESS).when(metadataManager).getLastProcessedLogEntryTimestamp();
+        logEntryWriter = new LogEntryWriter(corfuRuntime, replicationConfig, metadataManager);
+        txnContext = Mockito.mock(TxnContext.class);
+        Mockito.doReturn(txnContext).when(metadataManager).getTxnContext();
+    }
+
+    /**
+     * This test verifies that Log Entry(delta) updates are applied on the Sink with transaction boundary preserved.
+     * Each OpaqueEntry maps to a transaction on the Source.  Hence, each should be applied on the Sink as a separate
+     * transaction.
+     */
+    @Test
+    public void testLogEntryApplyWithExpectedTx() {
+        verifyNumberOfTx();
+    }
+
+    /**
+     * This test simulates the case that the same payload is received by LogEntryWriter twice and verifies that it is
+     * applied only once.
+     */
+    @Test
+    public void testLogEntryRedundantApply() {
+        verifyNumberOfTx();
+
+        int numOpaqueEntries = 3;
+        int topologyConfigId = 5;
+
+        // Send the same payload again
+        LogReplicationEntryMsg lrEntryMsg = generateLogEntryMsg(1, numOpaqueEntries, Address.NON_ADDRESS,
+            topologyConfigId);
+
+        txnContext = Mockito.mock(TxnContext.class);
+        Mockito.doReturn(txnContext).when(metadataManager).getTxnContext();
```

Comment:
Cleaned these tests and now using verifyNumberOfTx() for just the verification so reusing it again.  

txnContext must be reset so that it does not carry forward the previous invocations of commit().